src/vumi2/applications/turn_channels_api/messages.py

```diff
@@ -0,0 +1,196 @@
+      are supported.
+    """
+    ev = {
+        "id": event.user_message_id,
+        "timestamp": str(int(event.timestamp.timestamp())),
+        "status": None,
+    }
+    if event.event_type == EventType.DELIVERY_REPORT:
+        ev["status"] = {
+            DeliveryStatus.PENDING: "sent",
+            DeliveryStatus.FAILED: "sent",
+            DeliveryStatus.DELIVERED: "delivered",
+            None: None,
+        }[event.delivery_status]
+    else:
+        ev["status"] = "sent"
+    return ev
+
+
+@define
+class TurnOutboundMessage:
+    """
+    From
+    https://whatsapp.turn.io/docs/api/channel_api#receiving-outbound-messages-from-your-channel
+
+    * to (str) - the address (e.g. MSISDN) to send the message to.
+    * from (str) - the address the message is from. May be null if the
+      channel only supports a single from address.
+    * group (str) - If supported by the channel, the group to send the
+      messages to. Not required, and may be null
+    * reply_to (str) - the uuid of the message being replied to if this
+      is a response to a previous message. Important for session-based
+      transports like USSD. Optional.
+      The default settings allow 10 minutes to reply to a
+      message, after which an error will be returned.
+    * content (str) - The text content of the message. Required.
+    * priority (int) - Delivery priority from 1 to 5. Higher priority
+      messages are delivered first. If omitted, priority is 1. Not yet
+      implemented.
+    * channel_data (dict) - Additional data that is passed to the
+      channel to interpret. E.g. continue_session for USSD,
+      direct_message or tweet for Twitter.
+    """
+
+    content: str
+    to: str | None = None
+    from_addr: str | None = None
+    group: str | None = None
+    reply_to: str | None = None
+    priority: int = 1
+    channel_data: dict[str, Any] = field(factory=dict)
+
+    def __attrs_post_init__(self):
+        if self.to is None and self.reply_to is None:
+            raise ApiUsageError('Either "to" or "reply_to" must be specified')
+
+    @classmethod
+    def deserialise(
+        cls, data: dict[str, Any], default_from: str | None = None
+    ) -> "TurnOutboundMessage":
+        for key in ["context", "turn", "to"]:
+            if key not in data:
+                raise ApiUsageError(f"Missing key: {key}")
+        if "contact" not in data["context"]:
+            raise ApiUsageError("Missing key: context.contact")
+        if "text" not in data["turn"]:
+            raise ApiUsageError("Missing key: turn.text")
+        if "body" not in data["turn"]["text"]:
+            raise ApiUsageError("Missing key: turn.text.body")
+        contact = data["context"]["contact"]
+        return TurnOutboundMessage(
+            content=data["turn"]["text"]["body"],
+            to=data["to"],
+            from_addr=default_from,
+            group=contact["groups"][0]["name"],
```

Comment:
I think we can remove `group`, since we don't need it for USSD and SMS, and then we don't need to look in the `context` for now.